File: Examples/ObjectiveC/PointHeatmapExample.m

#import "PointHeatmapExample.h"
@import Mapbox;

NSString *const MBXExamplePointHeatmap = @"PointHeatmapExample";

@interface PointHeatmapExample () <MGLMapViewDelegate>

@end

@implementation PointHeatmapExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    MGLMapView *mapView = [[MGLMapView alloc] initWithFrame:self.view.frame styleURL:[MGLStyle darkStyleURLWithVersion:9]];
    mapView.delegate = self;
    
    [self.view addSubview:mapView];
}

- (void)mapView:(MGLMapView *)mapView didFinishLoadingStyle:(MGLStyle *)style {
    
   // Parse GeoJSON data. This example uses all M1.0+ earthquakes from 12/22/15 to 1/21/16 as logged by USGS' Earthquake hazards program.
        NSURL *url = [NSURL URLWithString:@"https://www.mapbox.com/mapbox-gl-js/assets/earthquakes.geojson"];

    MGLShapeSource *symbolSource = [[MGLShapeSource alloc] initWithIdentifier:@"symbol-source"];
    MGLSymbolStyleLayer *symbolLayer = [[MGLSymbolStyleLayer alloc] initWithIdentifier:@"place-city-sm" source:symbolSource];
    
    // Set the MGLShapeSourceOptions to allow clustering.
    NSDictionary *options = @{
                              MGLShapeSourceOptionClustered : @TRUE,
                              MGLShapeSourceOptionClusterRadius : @20,
                              MGLShapeSourceOptionMaximumZoomLevel : @15
                              };
    
    MGLShapeSource *earthquakeSource = [[MGLShapeSource alloc] initWithIdentifier:@"earthquakes" URL:url options:options];
    [style addSource:earthquakeSource];

    // Create a stops dictionary. The keys represent the number of points in a cluster.
    NSDictionary *stops = @{
                            @0: [MGLStyleValue valueWithRawValue:[UIColor yellowColor]],
                            @20.0: [MGLStyleValue valueWithRawValue:[UIColor orangeColor]],
                            @150.0: [MGLStyleValue valueWithRawValue:[UIColor redColor]],
                            };
    
    
    // Create and style the clustered circle layer.
    MGLCircleStyleLayer *clusteredLayer = [[MGLCircleStyleLayer alloc] initWithIdentifier:@"clustered layer" source:earthquakeSource];
    clusteredLayer.circleColor = [MGLStyleValue valueWithInterpolationMode:MGLInterpolationModeExponential
                                                sourceStops:stops
                                                attributeName:@"point_count"
                                                options:@{MGLStyleFunctionOptionDefaultValue: [MGLConstantStyleValue valueWithRawValue:[UIColor yellowColor]]}];
    
    clusteredLayer.circleRadius = [MGLConstantStyleValue valueWithRawValue:@70];
    clusteredLayer.circleOpacity = [MGLConstantStyleValue valueWithRawValue:@0.5];
    clusteredLayer.circleBlur = [MGLConstantStyleValue valueWithRawValue:@1];
    
    [style insertLayer:clusteredLayer belowLayer:symbolLayer];
}

@end

File: Examples/Swift/PointHeatmapExample.swift

import Mapbox

@objc(PointHeatmapExample_Swift)

class PointHeatmapExample_Swift: UIViewController, MGLMapViewDelegate {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let mapView = MGLMapView(frame: view.bounds)
        mapView.delegate = self
        
        view.addSubview(mapView)
    }
    
    func mapView(_ mapView: MGLMapView, didFinishLoading style: MGLStyle) {
        
        DispatchQueue.global().async {
            guard let url = URL(string: "https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_week.geojson") else { return }
            DispatchQueue.main.async {
                self.displayEarthquakes(url: url, style: style)
            }
        }
    }
    
    func displayEarthquakes(url: URL, style: MGLStyle) {
        let symbolSource = MGLSource(identifier: "symbol-source")
        let symbolLayer = MGLSymbolStyleLayer(identifier: "place-city-sm", source: symbolSource)
        
        //
        let options = [MGLShapeSourceOption.clustered: true,
                       MGLShapeSourceOption.clusterRadius: 20,
                       MGLShapeSourceOption.maximumZoomLevel: 15] as [MGLShapeSourceOption : Any]
        
        let earthquakeSource = MGLShapeSource(identifier: "earthquakes", url: url, options: options)
        style.addSource(earthquakeSource)
        
        let unclusteredLayer = MGLCircleStyleLayer(identifier: "unclustered", source: earthquakeSource)
        
        unclusteredLayer.circleColor = MGLConstantStyleValue(rawValue: UIColor(colorLiteralRed: 229/255.0, green: 94/255.0, blue: 94/255.0, alpha: 1))
        unclusteredLayer.circleRadius = MGLConstantStyleValue(rawValue: 20)
        unclusteredLayer.circleBlur = MGLConstantStyleValue(rawValue: 15)
        unclusteredLayer.predicate = NSPredicate(format: "%K != YES", argumentArray: ["cluster"])
        style.insertLayer(unclusteredLayer, below: symbolLayer)
        
        let stops = [
            0.0: MGLStyleValue(rawValue: UIColor(colorLiteralRed: 251/255.0, green: 176/255.0, blue: 59/255.0, alpha: 1)),
            20.0: MGLStyleValue(rawValue: UIColor(colorLiteralRed: 249/255.0, green: 136/255.0, blue: 108/255.0, alpha: 1)),
            150.0: MGLStyleValue(rawValue: UIColor(colorLiteralRed: 229/255.0, green: 94/255.0, blue: 94/255.0, alpha: 1))
        ]
        
        let circles = MGLCircleStyleLayer(identifier: "clustered layer", source: earthquakeSource)
        circles.circleColor = MGLStyleValue(interpolationMode: .exponential, sourceStops: stops, attributeName: "point_count", options: [.defaultValue: MGLStyleValue(rawValue: UIColor(colorLiteralRed: 251/255.0, green: 176/255.0, blue: 59/255.0, alpha: 1))])
        circles.circleRadius = MGLConstantStyleValue(rawValue: NSNumber(integerLiteral: 70))
        circles.circleBlur = MGLConstantStyleValue(rawValue: NSNumber(integerLiteral: 1))
        
        style.insertLayer(circles, below: symbolLayer)
    }
}
